### PR TITLE
feat: [AutoDeploy] unfusing attention

### PR DESCRIPTION
**Pull request by lucaslie**
This PR introduces several new features to support unfused attention in AutoDeploy:

- [x] More flexible factory interface for model, checkpoint loading, and tokenizer. Also cleans up weight loading and downloading logic to only download weights when necessary
- [x] Cleaned up and removed deprecated attention-related custom ops
- [x] Introducing the use of `position_ids` in addition to `input_ids` as standard input to the model. `position_ids` are necessary to correctly run inference when rope or other operators relying on `position_ids` cannot be pattern matched. Previously, we ignored `position_ids` since we assumed we can remove any position_ids-related op by op fusion.
- [x] Deprecated manually specified PoseConfig and AttentionInfo
- [x] Updated cached attention ops to _not_ to rope
- [x] Pattern matching to handle typical representation of attention in graphs: `repeat_kv`, `attn_mask`, `eager` attention, and `sdpa` attention
- [x] Separated cached attention ops from high-level pattern matching. Cached attention ops will now be inserted at the very end. Cached attention can now also infer cache shapes and other configs directly via appropriate fake tensors in node's meta dictionary.
- [x] Skipping attention pattern matching can now be achieved by simply specifying the corresponding `source_attention_op` that will be picked by the caching logic
- [x] Cleaned up shape propagation utilities so we can more reliably use FakeTensor meta data
- [x] Improved `torch_attention` reference ops
- [x] Clean up of transpose logic due to frequent bsnd<>bnsd switching
- [x] Unified cache insertion logic for MLA and MHA
- [x] Added unit tests where appropriate 
- [x] Fix remaining test failures
- [x] Handling attn_mask in `match_attention_layout` and add unit test for it
- [x] perf comparison between `main` and current commit

Some simple ballpark benchmarking:

`main`
```

===========================================================
= PERFORMANCE OVERVIEW 
===========================================================
Request Throughput (req/sec):                     21.6590
Total Output Throughput (tokens/sec):             2772.3489
Total Token Throughput (tokens/sec):              5544.6979
Total Latency (ms):                               46170.2345
Average request latency (ms):                     38228.2985
Per User Output Throughput [w/ ctx] (tps/user):   3.3901
Per GPU Output Throughput (tps/gpu):              2772.3489
```
vs 

`ll/unfused_sdpa`
```
===========================================================
= PERFORMANCE OVERVIEW 
===========================================================
Request Throughput (req/sec):                     21.3927
Total Output Throughput (tokens/sec):             2738.2698
Total Token Throughput (tokens/sec):              5476.5395
Total Latency (ms):                               46744.8467
Average request latency (ms):                     38936.7007
Per User Output Throughput [w/ ctx] (tps/user):   3.3297
Per GPU Output Throughput (tps/gpu):              2738.2698
```

### CODE REVIEW

File: examples/auto_deploy/simple_config.py [hunk 1]
```diff
@@ -39,7 +41,14 @@ class SimpleConfig:
         }
     )
 
-    ### CONFIGURE MODEL FACTORY, BACKEND, RUNTIME, AND WORLD SIZE ##################################
+    ### TOKENIZER EXTRA KWARGS #####################################################################
+    # Extra kwargs for the tokenizer class to customize the tokenizer. Same as model_kwargs.
+    # For example, the default HF Llama tokenizer can be initialized with the arguments specified
+    # [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/tokenization_llama_fast.py#L127).
+    # NOTE: This is only used if customize_tokenizer is True an
```
Comment (Fridah-nv) [inline examples/auto_deploy/simple_config.py]:
```suggestion
    # NOTE: This is only used if customize_tokenizer is True
```

File: tensorrt_llm/_torch/auto_deploy/models/hf.py [hunk 2]
```diff
@@ -47,58 +48,60 @@ def __init__(
...
-        hf_config: Dict[str, Any] = getattr(model, "config", object())
-
-        # TODO: let's see how we can support more Rope variants in the future
-        self._pos_embd_config = PositionalEmbeddingConfig(
-            mode="rope" if hasattr(hf_config, "rope_theta") else None,
-            rope_theta=getattr(hf_config, "rope_theta", 0.0),
-            rope_scale=1.0,
-        )
+        # patch forward method
+        model.forward = types.MethodType(self._simple_forward, model)
```
Comment (Fridah-nv) [inline tensorrt_llm/_torch/auto_deploy/models/hf.py:125]: Could you explain why it is needed to patch forward function with simple inputs?

File: tensorrt_llm/_torch/auto_deploy/transformations/library/rope.py [hunk 3]
```diff
@@ -335,10 +335,10 @@ def _process_rope_v1(
...
     with graph.inserting_before(q_match["add_node"]):
         if need_transpose:
-            q_for_op = graph.call_function(torch.ops.aten.transpose, args=(q_node, 1, 2))
-            k_for_op = graph.call_function(torch.ops.aten.transpose, args=(k_node, 1, 2))
-            q_for_op_contig = graph.call_method("contiguous", (q_for_op,))
-            k_for_op_contig = graph.call_method("contiguous", (k_for_op,))
+            q_for_op = graph.call_function(torch.ops.aten.transpose.int, args=(q_node, 1, 2))
+            k_for_op = graph.call_function(torch.ops.aten.transpose.int, args=(k_node, 1, 2))
+            q_for_op_contig = graph.call_function(torch.ops.aten.contiguous, args=(q_for_op,))
+            k_for_op_contig = graph.call_function(torch.ops.aten.contiguous, args=(k_for_op,))
```
Comment (Fridah-nv) [inline tensorrt_llm/_torch/auto_deploy/transformations/library/rope.py:341]: Thank you for the catch!

File: tensorrt_llm/_torch/auto_deploy/transformations/transform.py [hunk 4]
```diff
@@ -94,41 +102,52 @@ def __call__(self, cm: CachedSequenceInterface) -> GraphModule:
...
+        egm = match_grouped_attention(egm)
+
+        # Match and optimize causal attention masks
+        egm = match_causal_attn_mask(egm)
+
+        # Match attention layout expected by our backend
+        egm = match_attention_layout(egm, self.attention_op)
+
+        # Match rope
+        egm = match_rope_v1(egm)
```
Comment (Fridah-nv) [inline tensorrt_llm/_torch/auto_deploy/transformations/transform.py:123]: Should we move rope matching to fusion and optimization stage? Can also do in later PR though

File: tensorrt_llm/_torch/auto_deploy/transformations/_graph.py [hunk 5]
```diff
@@ -132,9 +135,24 @@ def canonicalize_graph(gm: GraphModule, shape_prop: bool = False) -> GraphModule
 
     # NOTE: shape_prop can be a littly finicky & slow, so we only run it optionally...
     if shape_prop:
-        inps = tuple([node.meta.get("val") for node in gm.graph.nodes if node.op == "placeholder"])
-        with lift_to_meta(gm):
-            FakeTensorProp(gm, _detect_fake_mode_from_gm(gm)).run(*inps)
+        fake_mode: Optional[FakeTensorMode] = _detect_fake_mode_from_gm(gm)
```
Comment (Fridah-nv) [inline tensorrt_llm/_torch/auto_deploy/transformations/_graph.py:163]: Please comment on why `args_static` needs special handling and the example use case

File: tensorrt_llm/_torch/auto_deploy/custom_ops/flashinfer_attention.py [hunk 6]
```diff
@@ -20,19 +35,10 @@ class PlanParams:
     page_size: int
     q_dtype: torch.dtype
     kv_dtype: torch.dtype
+    sm_scale: Optional[float] = None
```
Comment (Fridah-nv) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/flashinfer_attention.py:38]: Does the separation of rope and attention expose softmax_scale so it needs to be added here?

**Comment by lucaslie** [comment]
/bot run --disable-fail-fast --stage-list "DGX_H100-4_GPUs-PyTorch-[Post-Merge]"

File: tensorrt_llm/_torch/auto_deploy/custom_ops/flashinfer_attention.py (hunk 6 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/flashinfer_attention.py:38]: Yes, that's the case. We now need to carry mask and scale information around

File: tensorrt_llm/_torch/auto_deploy/transformations/transform.py (hunk 4 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/transformations/transform.py:123]: Let's do this later. Right now, this would trip up the TP sharder. Once you have a chance to update the TP sharder we can revisit this. I will leave a todo in the meantime

File: tensorrt_llm/_torch/auto_deploy/transformations/_graph.py (hunk 5 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/transformations/_graph.py:163]: added a better docstring

File: tensorrt_llm/_torch/auto_deploy/models/hf.py (hunk 2 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/models/hf.py:125]: updated the docstring in `factory.py` to explain the requirements

File: tensorrt_llm/_torch/auto_deploy/transformations/transform.py (hunk 4 above)
Comment (Fridah-nv) [reply tensorrt_llm/_torch/auto_deploy/transformations/transform.py:123]: Got it, thanks!

**Review by suyoggupta — APPROVED** [review]

**Comment by lucaslie** [comment]
TODO: let's look at perf numbers for `trtllm-bench` similar to @suyoggupta's numbers

**Comment by lucaslie** [comment]
/bot run --disable-fail-fast --extra-stage "DGX_H100-4_GPUs-PyTorch-[Post-Merge]"

**Comment by lucaslie** [comment]
/bot run --disable-fail-fast --extra-stage "DGX_H100-4_GPUs-PyTorch-[Post-Merge]"
